lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2569,141 +2580,213 @@ buildAndSignTransactionNew
+    argGenChange era preSelection txCtx = do
+    --
+    stdGen <- initStdGen
+    pureTimeInterpreter <- snapshot ti
+    protocolParameters <- currentProtocolParameters netLayer
+    let wrapError = ExceptionWitnessTx . ErrWitnessTxWithRootKey
+    throwOnError <=< runExceptT $ withRootKey db walletId pwd wrapError $
+        \rootKey scheme -> lift $ atomically $ do
+            pendingTxs <- fmap fromTransactionInfo <$>
+                readTransactions
+                    walletId Nothing Descending wholeRange (Just Pending)
+            txBuiltForSlot <- modifyDBMaybe walletsDB $ do
+                let err = ExceptionConstructTx . ErrConstructTxNoSuchWallet
+                adjustNoSuchWallet walletId err $ \(s :: WalletState s) -> do
+                    let wallet = WalletState.getLatest s
+                    let (errorOrBuiltTx, wallet') =
+                            buildAndSignTransactionPure @k @ktype @s @n
+                                nullTracer
+                                pureTimeInterpreter
+                                (Set.fromList pendingTxs)
+                                rootKey
+                                scheme
+                                pwd
+                                protocolParameters
+                                txLayer
+                                argGenChange
+                                era
+                                preSelection
+                                txCtx
+                            & runExceptT . withExceptT
+                                (either ExceptionBalanceTx ExceptionConstructTx)
+                            & (`runStateT` wallet)
+                            & (`evalRand` stdGen)
+
+                    -- Newly generated change addresses only change the Prologue
+                    Right ( [ReplacePrologue $ getPrologue $ getState wallet']
+                          , (errorOrBuiltTx, currentTip wallet ^. #slotNo)
+                          )
+
+            let slotTime slot = liftIO $ interpretQuery
+                    (neverFails "slot is ahead of the node tip" ti)
+                    (slotToUTCTime slot)
+
+            bitraverse throwOnError slotTime =<< throwOnError txBuiltForSlot
+  where
+    throwOnError :: (MonadIO m, Exception e) => Either e a -> m a
+    throwOnError = either (liftIO . throwIO) pure
+
+data BuiltTx = BuiltTx
+    { builtTx :: Tx
+    , builtTxMeta :: TxMeta
+    , builtSealedTx :: SealedTx
+    }
+
+buildAndSignTransactionPure
+    :: forall k ktype s (n :: NetworkDiscriminant) m
+     . ( Typeable n
+       , Typeable s
+       , Typeable k
+       , WalletKey k
+       , HardDerivation k
+       , BoundedAddressLength k
+       , Bounded (Index (AddressIndexDerivationType k) (AddressCredential k))
+       , IsOwned s k ktype
+       , IsOurs s RewardAccount
+       , GenChange s
+       , MonadRandom m
+       , m ~ StateT (Wallet s) (Rand StdGen)
```

Comment:
Parameterizing by `m` but then having this constraint seems deceptive to me. What do you think about having the  `StateT (Wallet s) (Rand StdGen)` in the return-type directly?

Also, I'm guessing there might be no `Tracer  (StateT (Wallet s) (Rand StdGen)) WalletLog` values besides `nullTracer` making the argument pointless at the moment. It might be better to remove then, unless you plan on allowing other choices of `m`.

Reply:
Good point on removing unused tracer, in which case `m` could as well be inlined, thanks!